aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -265,51 +236,103 @@ Tensor & masked_select_out_cuda(Tensor & result, const Tensor & self, const Tens
+      cuda_take_put_kernel<scalar_t, index_t>(sub_iter, indexed, f);
+    }
+    return;
+  }
+
+  auto indexed_info = cuda::detail::getTensorInfo<scalar_t, index_t>(indexed);
 
-  TORCH_CHECK(output.layout() == Layout::Strided, "take() only supports strided layout, got layout: ", output.layout(), " on output tensor");
-  TORCH_CHECK(input.layout() == Layout::Strided, "take() only supports strided layout, got layout: ", input.layout(), " on input tensor");
-  TORCH_CHECK(index.layout() == Layout::Strided, "take() only supports strided layout, got layout: ", index.layout(), " on index tensor");
+  const auto numel = indexed.numel();
+  const bool is_contiguous = indexed.is_contiguous();
 
-  TORCH_CHECK(output.scalar_type() == input.scalar_type(),
-          "output and input scalar type must match. but got different types: ", output.scalar_type(), " and ", input.scalar_type());
-  TORCH_CHECK(index.scalar_type() == kLong, "index must be an int64 tensor");
+  char* __restrict__ iterated_ptr = reinterpret_cast<char*>(iter.data_ptr(0));
+  char* __restrict__ idx_ptr = reinterpret_cast<char*>(iter.data_ptr(1));
 
-  TensorArg output_arg{ output, "output", 1 };
-  TensorArg input_arg{ input, "input", 2 };
-  TensorArg index_arg{ index, "index", 3 };
-  checkAllSameGPU("take", {output_arg, input_arg, index_arg});
+  const auto offset_calc = make_offset_calculator<2>(iter);
 
-  TORCH_CHECK(input.dim() < MAX_CUTORCH_DIMS, CUTORCH_DIM_WARNING);
-  TORCH_CHECK(output.dim() < MAX_CUTORCH_DIMS, CUTORCH_DIM_WARNING);
-  TORCH_CHECK(index.dim() < MAX_CUTORCH_DIMS, CUTORCH_DIM_WARNING);
+  auto loop = [=]C10_DEVICE(int i) {
+    auto offsets = offset_calc.get(i);
 
-  TORCH_CHECK(!(input.numel() == 0 && index.numel() != 0), "tried to take from an empty tensor");
+    auto& iterated = *reinterpret_cast<scalar_t*>(iterated_ptr + offsets[0]);
+    const auto idx = *reinterpret_cast<int64_t*>(idx_ptr + offsets[1]);
+    CUDA_KERNEL_ASSERT(idx < numel && idx >= -numel && "cuda_take_put_kernel() index out of bounds");
+    const auto offset = indexToOffset<scalar_t, index_t>(indexed_info, idx, numel, is_contiguous);
 
-  at::assert_no_internal_overlap(output);
-  at::assert_no_partial_overlap(output, index);
-  at::assert_no_overlap(output, input);
+    f(iterated, indexed_info.data, offset);
+  };
+  launch_kernel<launch_size_nd, launch_bound2>(iter.numel(), loop);
+}
 
-  output.resize_(index.sizes());
 
-  AT_DISPATCH_ALL_TYPES_AND2(at::ScalarType::Bool, at::ScalarType::Half, input.scalar_type(), "take_cuda", [&] {
-    dispatchTakePut<scalar_t>(input, output, index);
+void put_kernel(
+  TensorIterator& iter,
+  const Tensor& output,
+  const bool accumulate) {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(at::ScalarType::Half, at::ScalarType::Bool, at::ScalarType::BFloat16, iter.dtype(), "put_cuda", [&] {
+    // Cannot use `OpaqueType`, as we need the actual type for `fastSpecializedgpuAtomicAdd` and `TensorInfo`
```

Comment:
you can use AT_DISPATCH_INDEX_TYPE here to avoid 2 similar codeblocks.

Reply:
Could you elaborate on this? The index may only be of type `int64` due to this line
https://github.com/pytorch/pytorch/blob/91fd8d571a54a46155a79d21cd0fcd6c1effe725/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L322
This code is supposed to handle the case when `output` is too large, since `output` is not iterated over with the tensor iterator. That being said, I am not completely what it does it correct, so I would appreciate some feedback on this. It is done as it was done in the previous code, but I am not 100% sure that what it does is indeed correct.

Reply:
AT_DISPATCH_INDEX_TYPES defines index_t to either int32_t or int64_t, pretty much like you are doing here, here's an example of use: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L476-L482. It doesn't require that actual index is of index_t, it's up to you how you write you lambda that that dispatch macro calls. It's just a convenience for what you are doing manually here.

Reply:
Sorry, but I am not completely sure I am following. Do you suggest I write the following:
```c++
if (cuda::detail::canUse32BitIndexMath(output)) {
  AT_DISPATCH_INDEX_TYPES(int32_t, "put_cuda_int32",
    [&] {
       if (accumulate) {
         const auto numel = output.numel();
         cuda_take_put_kernel<scalar_t, index_t>(iter, output,
               [numel] __device__(scalar_t& iterated, scalar_t* __restrict__ indexed, const index_t offset) {
                 fastSpecializedAtomicAdd(indexed, offset, numel, iterated);
               });
       }
       else {
         cuda_take_put_kernel<scalar_t, index_t>(iter, output,
              [] __device__(scalar_t& iterated, scalar_t* __restrict__ indexed, const index_t offset) {
                 indexed[offset] = iterated;
               });
       }
    });
  }
else {
  AT_DISPATCH_INDEX_TYPES(int64_t, "put_cuda_int64",
    [&] {
       if (accumulate) {
         const auto numel = output.numel();
         cuda_take_put_kernel<scalar_t, index_t>(iter, output,
               [numel] __device__(scalar_t& iterated, scalar_t* __restrict__ indexed, const index_t offset) {
                 fastSpecializedAtomicAdd(indexed, offset, numel, iterated);
               });
       }
       else {
         cuda_take_put_kernel<scalar_t, index_t>(iter, output,
              [] __device__(scalar_t& iterated, scalar_t* __restrict__ indexed, const index_t offset) {
                 indexed[offset] = iterated;
               });
       }
    });
}
```
I think I am missing something, as this code is pretty much the same as the one that is there at the moment.

Reply:
In fact, this is the best I could come up with, but I still do not see how to use `AT_DISPATCH_INDEX_TYPES` to simplify the code...
```c++
template<typename scalar_t, typename index_t>
void put_cuda_impl(
  TensorIterator& iter,
  const Tensor& output,
  const bool accumulate) {
   if (accumulate) {
     const auto numel = output.numel();
     cuda_take_put_kernel<scalar_t, index_t>(iter, output,
           [numel] __device__(scalar_t& iterated, scalar_t* __restrict__ indexed, const index_t offset) {
             fastSpecializedAtomicAdd(indexed, offset, numel, iterated);
           });
   }
   else {
     cuda_take_put_kernel<scalar_t, index_t>(iter, output,
          [] __device__(scalar_t& iterated, scalar_t* __restrict__ indexed, const index_t offset) {
             indexed[offset] = iterated;
           });
   }
}

void put_kernel(
  TensorIterator& iter,
  const Tensor& output,
  const bool accumulate) {
  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(at::ScalarType::Half, at::ScalarType::Bool, at::ScalarType::BFloat16, iter.dtype(), "put_cuda", [&] {                                                                                             299     // Cannot use `OpaqueType`, as we need the actual type for `fastSpecializedgpuAtomicAdd` and `TensorInfo`
    if (cuda::detail::canUse32BitIndexMath(output)) {
      put_cuda_impl<scalar_t, int32_t>(iter, output, accumulate);
    }
    else {
      put_cuda_impl<scalar_t, int64_t>(iter, output, accumulate);
    }
  });
}
```

Reply:
OH Nevermind, I just understood what you meant, sorry. I am still getting used to having "types that are not static"!